syntax = "proto3";
package RevaProtocol;
import "RevaVariable.proto";
option java_package = "reva.protocol";

service RevaDecompilationService {
    rpc GetDecompilation(RevaGetDecompilationRequest) returns (RevaGetDecompilationResponse) {}
    rpc RenameFunctionVariable(RevaRenameFunctionVariableRequest) returns (RevaRenameFunctionVariableResponse) {}
    rpc GetFunctionList(RevaGetFunctionListRequest) returns (stream RevaGetFunctionListResponse) {}
}

message RevaGetDecompilationRequest {
    string function = 1;
    string address = 2;
}

message RevaGetDecompilationResponse {
    uint64 address = 1;
    string function = 2;
    string function_signature = 3;
    repeated string incoming_calls = 4;
    repeated string outgoing_calls = 5;
    repeated RevaVariable variables = 6;
    string listing = 7;
    string decompilation = 8;
    string error_message = 9;
}

message RevaRenameFunctionVariableRequest {
    string function_name = 1;
    string old_name = 2;
    string new_name = 3;
}

message RevaRenameFunctionVariableResponse {
}

message RevaGetFunctionListRequest {
}

message RevaGetFunctionListResponse {
    string function_name = 1;
    string function_signature = 2;
    string entry_point = 3;
    repeated string incoming_calls = 4;
    repeated string outgoing_calls = 5;
}